Do not clear refs on recursive type references

peterbroadhurst opened the pull request:
Proposed UT and fix for #447 

A reference was being added carefully in the recursion checker while building the struct on this line (note this _also_ includes the `schema` being passed into the `value` of the `$ref` - we leave both):
https://github.com/getkin/kin-openapi/blob/0db2d4c3583875f4fe65b47ccb51662a20ed32a9/openapi3gen/openapi3gen.go#L344

However, it was then being stripped out in this code:
https://github.com/getkin/kin-openapi/blob/0db2d4c3583875f4fe65b47ccb51662a20ed32a9/openapi3gen/openapi3gen.go#L58-L61

As the original `$ref.value` was still there (see note above), this meant the infinite loop occurred on JSON serialization.

peterbroadhurst commented:
Converting to draft while I work with @nguyer to verify this in the overall context of the Swagger generation.

fenollp commented:
Hi! Thanks for the patch. Before reviewing I'd like to note that this lib is v0 so there should not be much care given to backwards compatibility especially in this very case where the default behavior is troublesome - does not bode well with recursive refs.
Could you adapt your patch to actually replace the previous implementation namely `NewSchemaRefForValue`?

peterbroadhurst commented:
I've made the update to replace `NewSchemaRefForValue`, which did affect some tests so:

- I've updated tests that wanted to inspect internal details, to use `g := NewGenerator()` directly and call the internal implementation method
- As there was an assertion in the `simple_test.go` that could no longer be there, and that I believe is really more of a sample, I added a test that verifies the sample behavior.

nguyer commented:
Thanks @peterbroadhurst! I just tested this branch and these changes fix the issue I described in https://github.com/getkin/kin-openapi/issues/447

fenollp commented:
Superseeded by https://github.com/getkin/kin-openapi/pull/454
Please open an issue / a PR if there's anything you want working differently.
